### PR TITLE
teleport KSM token between Kusama and Basilisk

### LINKED ISSUE #4035: Implement teleport KSM token from Kusama to Basilisk
### Description

I am just wondering if teleport works for Basilisk :)

### What should be added.

We should have handlers for multicurrencies?

### Misc?

https://github.com/paraspell/sdk/issues/1

### PR DESCRIPTION
**Pull request by Jarsen136**
**Thank you for your contribution** to the [KodaDot NFT gallery](https://kodadot.xyz).

👇 \_ Let's make a quick check before the contribution.

## PR Type

- [ ] Bugfix
- [x] Feature
- [ ] Refactoring

## Context

- [x] Closes #4035 
- [ ] Requires deployment <>

#### Before submitting pull request, please make sure:

- [x] My contribution builds **clean without any errors or warnings**
- [x] I've merged recent default branch -- **main** and I've no conflicts
- [x] I've tried to respect high code quality standards
- [x] I've didn't break any original functionality
- [x] I've posted a screenshot of demonstrated change in this PR

#### Optional

- [ ] I've tested it at </bsx/collection>
- [ ] I've tested PR on mobile
- [ ] I've written unit tests 🧪
- [ ] I've found edge cases

#### Had issue bounty label?

- [x] Fill up your KSM address: [Payout](https://beta.kodadot.xyz/transfer/?target=Caiv9TbPz68q5dC8EcHu5xKYPRnremimGzqmEejDFNpWWLG)

#### Community participation

- [x] [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)

## Screenshot 📸

- [ ] My fix has changed **something** on UI; a screenshot is best to understand changes for others.

currently, the url for teleport is `/teleport-bridge`

### CODE REVIEW

**Comment by Jarsen136** [comment]
Work in progress:

- [x] setup teleport page on kodadot, according to https://github.com/paraspell/ui
- [x] teleport ksm token from Kusama to Basilisk

image

I have send 1 ksm from Kusama to Basilisk on teleport page. https://kusama.subscan.io/xcm_message/kusama-350df3d577486e2054b580f0065c872b4b4d693b
Still waiting for the confirmation at Basilisk network.

image



- [ ] teleport ksm token from Basilisk to Kusama
- [ ] refactor code of teleport component 
- [ ] UI redesign [teleport ](https://github.com/kodadot/nft-gallery/issues/4119)

**Comment by yangwao** [comment]
> Still waiting for the confirmation at Basilisk network.

Basilisk is bricked for now and won't work for like 5-6 days for sure :| 

BUT we can test it against Rococo at SNEK (Basilisk testnet), probably @vikiival can send you some testnet tokens there

**Comment by Jarsen136** [comment]
> BUT we can test it against
> 
> * Statemine (live net)
> * Rococo at SNEK (Basilisk testnet), probably @vikiival can send you some testnet tokens there

Oh, that would be better.
My Basilisk test address is bXk6PXHj9NnYkiZ45ETrq4iLiozW8a4khe8ZmqDkdbLJNNMaE

**Comment by yangwao** [comment]
Once merge, we can test this one 
- #4326

**Comment by yangwao** [comment]
It should working now? 👀

**Comment by Jarsen136** [comment]
Teleport bridge path: https://deploy-preview-4297--koda-nuxt.netlify.app/teleport-bridge

- [x] setup teleport page on kodadot, according to https://github.com/paraspell/ui
- [x] refactor code of teleport component
---
- [x] teleport ksm token from Kusama to Basilisk (✅ Tested)

Teleport Steps:
1. switch to rmrk network
2. input the bsx recipient address
3. send transaction
image

---
- [x] teleport ksm token from Basilisk to Kusama

Teleport Steps:
1. switch to bsx network
2. input the ksm recipient address
3. send transaction

I have tried many times to test the functionality that on Basilisk. However, it's hard for me to connect with the Basilisk websocket network. 

image

Could anyone do a teleport test from Kusama to Basilisk?

File: components/teleport/Teleport.vue
Comment (Jarsen136) [reply components/teleport/Teleport.vue:70]: Yes, I will transform it into a vue3 component.

**Comment by Jarsen136** [comment]
> 
> * [ ]  teleport ksm token from Basilisk to Kusama
> 
> Teleport Steps:
> 
> 1. switch to bsx network
> 2. input the ksm recipient address
> 3. send transaction
> 
> I have tried many times to test the functionality that on Basilisk. However, it's hard for me to connect with the Basilisk websocket network.
> 

I have found out why it does not work properly when I teleport ksm from Basilisk.

As we could see at `@paraspell/sdk` here : https://github.com/paraspell/sdk/blob/7fc8b1e87033ac4e0de024f81d52f6558474531e/src/pallets/xTokens/transfer.ts#L5

Function `transferParaToRelay` does not prepare the result for `Basilisk`.
image

So, I guess we should fulfill the function `transferParaToRelay ` for Basilisk first. 🤔 WDYT @vikiival

File: package.json
Comment (vikiival) [inline package.json:22]: ??

**Comment by vikiival** [comment]
> Function transferParaToRelay does not prepare the result for Basilisk.

@Dudo50

File: package.json
Comment (Jarsen136) [reply package.json:22]: It has been recovered

**Comment by Jarsen136** [comment]
✅  teleport ksm token from Basilisk to Kusama
It works! 
However, the network on Basilisk node seems not that stable.

image

**Comment by vikiival** [comment]
> However, the network on Basilisk node seems not that stable.

what does that mean?

**Comment by Jarsen136** [comment]
> > However, the network on Basilisk node seems not that stable.
> 
> what does that mean?

I have to wait for a long time to connect to Basilisk rpc node when I want to teleport token. I'm not sure it's related to my own network or not. It really kills my time : )
image

**Comment by dudo50** [comment]
@Jarsen136 Transfer Basilisk to Relay chain does not work because you use wrong version of sdk, currently version in main branch is suited for localhost use with UI only. However if you switch to beta-pre-release branch I think you will find what you were searching for :)..
If you have any questions feel free to contact me or tag me here.

**Comment by Jarsen136** [comment]
> @Jarsen136 Transfer Basilisk to Relay chain does not work because you use wrong version of sdk, currently version in main branch is suited for localhost use with UI only. However if you switch to beta-pre-release branch I think you will find what you were searching for :).. If you have any questions feel free to contact me or tag me here.

Hey, @dudo50 thanks for your answer. The `beta-pre-release` branch would help to transfer tokens from Basilisk. However, the latest version 0.0.11 for SDK is built from the main branch. So, I guess there is no specific released version of SDK that could support transferring tokens from Basilisk, for now. Correct me if i'm wrong.


For now, I write the transfer function call code without using any SDK. If the latest version of @paraspell/sdk is released in the future, I would switch to using the SDK to support transfer tokens on more network

**Comment by dudo50** [comment]
@Jarsen136 there is actually it is called 0.0.11.alpha I believe, you can check it on npm.. It supports Basilisk and Kusama transfers out of the box. However I respect your decision, I would like to encourage you to check it anyways so you can see how to construct calls you seek if there is something you still need to implement :)

**Comment by Jarsen136** [comment]
> @Jarsen136 there is actually it is called 0.0.11.alpha I believe, you can check it on npm.. It supports Basilisk and Kusama transfers out of the box. However I respect your decision, I would like to encourage you to check it anyways so you can see how to construct calls you seek if there is something you still need to implement :)

Thanks for correcting me : )  I will use this alpha version of SDK to help to build the teleport function. I think it would do better.

**Comment by dudo50** [comment]
No worries @Jarsen136 , it is still under construction so some things might not be clear right away. We plan on converting it into a builder patter soon.. If you have any questions feel free to tag me :)..

**Comment by Jarsen136** [comment]
After many times of testing, I think the functionality is good for testing now. @yangwao 

image

**Comment by yangwao** [comment]
i was randomly checking this pr, when was draft and suddenly green! 
can test and let you know

**Comment by yangwao** [comment]
@Jarsen136 maybe adding teleport-bridge to dropdown?

I would add it under Transfer for now :)
image

**Comment by Jarsen136** [comment]
> @Jarsen136, maybe adding teleport-bridge to dropdown?
> 
> I would add it under Transfer for now :) 

How about naming `Teleport Bridge`?
image

File: components/teleport/Teleport.vue
Comment (vikiival) [inline components/teleport/Teleport.vue:28]: AccountInput cries in the corner

File: components/teleport/Teleport.vue
Comment (vikiival) [inline components/teleport/Teleport.vue:47]: BalanceInput cries in the corner

File: components/teleport/Teleport.vue
Comment (vikiival) [inline components/teleport/Teleport.vue:56]: SubmitButton cries in third corner

File: components/teleport/Teleport.vue
Comment (vikiival) [inline components/teleport/Teleport.vue:77]: Plese use `sub-api` :)

File: components/teleport/Teleport.vue
Comment (vikiival) [inline components/teleport/Teleport.vue:142]: `transactionExecutor.ts`

**Review by vikiival — CHANGES_REQUESTED** [review]
Please use already made components

File: components/teleport/Teleport.vue
Comment (yangwao) [reply components/teleport/Teleport.vue:47]: BalanceInput

File: components/teleport/Teleport.vue
Comment (Jarsen136) [reply components/teleport/Teleport.vue]: done

File: components/teleport/Teleport.vue
Comment (Jarsen136) [reply components/teleport/Teleport.vue]: done

File: components/teleport/Teleport.vue
Comment (Jarsen136) [reply components/teleport/Teleport.vue]: done

File: components/teleport/Teleport.vue
Comment (Jarsen136) [reply components/teleport/Teleport.vue:73]: I'm afraid that it could not use `sub-api`. Because If I want to teleport tokens from Kusama, I need to use `ApiPromise` of kusama. On the other hand, I would use `ApiPromise` of basilisk. 

And `sub-api` would only return the API endpoint according to the current prefix.  👀

Comment (yangwao) [reply components/teleport/Teleport.vue:73]: https://github.com/kodadot/packages/issues/78

**Comment by yangwao** [comment]
I don't have my keys with me rn, so I believe it has worked! 

pay 150 USD

**Comment by yangwao** [comment]
😍 Perfect, I’ve sent the payout 
💵 $150 @ 25.92 USD/KSM ~ 5.787 $KSM
🧗 [Caiv9TbPz68q5dC8EcHu5xKYPRnremimGzqmEejDFNpWWLG](https://kusama.subscan.io/account/Caiv9TbPz68q5dC8EcHu5xKYPRnremimGzqmEejDFNpWWLG)
🔗 [0x09332075fa27f7301474f43751818f0e80f7518a5cea7bb46fdc8a85e3909a6a](https://kusama.subscan.io/extrinsic/0x09332075fa27f7301474f43751818f0e80f7518a5cea7bb46fdc8a85e3909a6a)

🪅 Let’s grab another issue and get rewarded! 
🪄 [github.com/kodadot/nft-gallery/issues](https://github.com/kodadot/nft-gallery/issues/)

**Comment by yangwao** [comment]
Let's now crack on #4119

**Pull request merged**
